optuna/trial/_frozen.py

```diff
@@ -321,8 +322,15 @@ def _validate(self) -> None:
                     "`datetime_complete` is supposed to be None for an unfinished trial."
                 )
 
+        if self.state == TrialState.PRUNED and self._values is not None:
+            raise ValueError("`value` is supposed not to be set for a pruned trial.")
+        if self.state == TrialState.FAIL and self._values is not None:
+            raise ValueError("`value` is supposed not to be set for a failed trial.")
         if self.state == TrialState.COMPLETE and self._values is None:
             raise ValueError("`value` is supposed to be set for a complete trial.")
+        self_values: List[float] = self._values if self._values is not None else []  # for mypy
+        if self.state == TrialState.COMPLETE and any([math.isnan(x) for x in self_values]):
+            raise ValueError("`value` is supposed not to be NaN for a complete trial.")
```

Comment:
```suggestion
        if self.state == TrialState.COMPLETE:
            if self._values is None:
                raise ValueError("values should be set for a complete trial.")
            elif any(math.isnan(x) for x in self._values):
                raise ValueError("values should not contain NaN.")
```
A better way to silence mypy